llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -352,124 +355,81 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
+      if (VisitedFunctions.find(ParentF) != VisitedFunctions.end())
         continue;
 
-      bool IsIndirectlyCalled =
-          HasIndirectlyCalledAttr ||
-          std::find(isIndirectlyCalledInGraph.begin(),
-                    isIndirectlyCalledInGraph.end(),
-                    CF) != isIndirectlyCalledInGraph.end();
-
-      // Return if we've already discovered if there are asserts in the
-      // function call graph.
-      auto HasAssert = hasAssertionInCallGraphMap.find(CF);
-      if (HasAssert != hasAssertionInCallGraphMap.end()) {
-        // If we know, that this function does not contain assert, we still
-        // should investigate another instructions in the function.
-        if (!HasAssert->second)
-          continue;
-
-        return IsIndirectlyCalled ? Assert_Indirect : Assert;
+      if (ParentF->hasFnAttribute("referenced-indirectly")) {
+        return {true, {}};
       }
 
-      if (CF->getName().startswith("__devicelib_assert_fail")) {
-        // Mark all the functions above in call graph as ones that can call
-        // assert.
-        for (const auto *It : FuncCallStack)
-          hasAssertionInCallGraphMap[It] = true;
-
-        hasAssertionInCallGraphMap[Func] = true;
-        hasAssertionInCallGraphMap[CF] = true;
-
-        return IsIndirectlyCalled ? Assert_Indirect : Assert;
-      }
+      if (ParentF->getCallingConv() == CallingConv::SPIR_KERNEL)
+        KernelNames.push_back(ParentF->getName());
 
-      if (!CF->isDeclaration()) {
-        Workstack.push_back(CF);
-        IsLeaf = false;
-        if (HasIndirectlyCalledAttr)
-          isIndirectlyCalledInGraph.push_back(CF);
-      }
-    }
-
-    if (IsLeaf && !FuncCallStack.empty()) {
-      // Mark the leaf function as one that definetely does not call assert.
-      hasAssertionInCallGraphMap[FuncCallStack.back()] = false;
-      FuncCallStack.clear();
+      FunctionsToVisit.push(ParentF);
     }
   }
-  return No_Assert;
+
+  return {false, std::move(KernelNames)};
 }
 
 std::vector<StringRef> getKernelNamesUsingAssert(const Module &M) {
-  std::vector<StringRef> Result;
-
-  bool HasIndirectlyCalledAssert = false;
-  EntryPointGroup Kernels;
-  for (const auto &F : M.functions()) {
-    // TODO: handle SYCL_EXTERNAL functions for dynamic linkage.
-    // TODO: handle function pointers.
-    if (F.getCallingConv() != CallingConv::SPIR_KERNEL)
-      continue;
-
-    Kernels.push_back(&F);
-    if (HasIndirectlyCalledAssert)
-      continue;
-
-    HasAssertStatus HasAssert = hasAssertInFunctionCallGraph(&F);
-    switch (HasAssert) {
-    case Assert:
-      Result.push_back(F.getName());
-      break;
-    case Assert_Indirect:
-      HasIndirectlyCalledAssert = true;
-      break;
-    case No_Assert:
-      break;
+  Optional<const Function *> DevicelibAssertFailFunction;
+  std::vector<StringRef> SPIRKernelNames;
+  // This loop finds all SPIR kernel's names and __devicelib_assert_fail
+  // function if it is present.
```

Comment:
I suggest we optimize this function:

```c++
auto DevicelibAssertFailFunction = M.getFunction("__devicelib_assert_fail");
if (!DevicelibAssertFailFunction)
  return {};

auto TraverseResult =
      TraverseCGToFindSPIRKernels(*DevicelibAssertFailFunction);
// Here I would return `true` if no "referenced-indirectly" functions are reached to make code more readable
if (TraverseResult.first)
  return TraverseResult.second;

// Here we reached "referenced-indirectly", so we need to find all kernels and return them.
std::vector<StringRef> SPIRKernelNames;
for (const Function &F : M) {
  if (F.getCallingConv() == CallingConv::SPIR_KERNEL)
    SPIRKernelNames.push_back(F.getName());
}
return SPIRKernelNames;
```

This way to go over all function **only if** there are indirectly callable functions invoking assert function. This should be a rare case.